File: Lab/SEM4/MPP/lab5/moto_java/src/main/java/org/zdroba/service/AuthServiceImpl.java
package org.zdroba.service;

import org.mindrot.jbcrypt.BCrypt;
import org.zdroba.entity.User;
import org.zdroba.exceptions.AlreadyExistsException;
import org.zdroba.exceptions.InvalidPasswordException;
import org.zdroba.exceptions.NotFoundException;
import org.zdroba.repository.UserRepository;
import org.zdroba.repository.UserRepositoryImpl;

public class AuthServiceImpl implements AuthService{

    private final UserRepository userRepository;

    public AuthServiceImpl(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    public User logIn(String email, String password) throws NotFoundException, InvalidPasswordException {
        User user = userRepository.find(email);

        if(user == null)
            throw new NotFoundException(String.format("User with email: %s does not exist", email));

        if(!BCrypt.checkpw(password, user.getPassword()))
            throw new InvalidPasswordException("Passwords do not match");

        return user;
    }

    @Override
    public User register(String email, String password) throws AlreadyExistsException {
        User user = userRepository.find(email);

        if(user != null)
            throw new AlreadyExistsException(String.format("User with email: %s does exists", email));

        user = new User(email, BCrypt.hashpw(password, BCrypt.gensalt()));
        userRepository.add(user); // in repo, .add mutates the object by adding generated id

        return user;
    }
}

File: Lab/SEM4/MPP/lab8/moto_c/src/Client/service/AuthServiceImpl.cs
using moto_c.entity;
using moto_c.exceptions;
using moto_c.repository;
using BCrypt.Net;   

namespace moto_c.service;

public class AuthServiceImpl:AuthService
{
    private UserRepository userRepository;

    public AuthServiceImpl(UserRepository userRepository)
    {
        this.userRepository = userRepository;
    }

    public User? logIn(string email, string password)
    {
        User? user = userRepository.find(email);

        if (user == null)
            throw new NotFoundException("User not found");
        
        if(!BCrypt.Net.BCrypt.Verify(password, user.password))
            throw new InvalidPasswordException("Password is incorrect");

        return user;
    }

    public User? register(string email, string password)
    {
        User? user = userRepository.find(email);

        if (user != null)
            throw new AlreadyExistsException("User with this email already exists");

        user = new User(email, BCrypt.Net.BCrypt.HashPassword(password));
        userRepository.add(user);

        return user;
    }
}